frame/support/src/lib.rs

```diff
@@ -946,3 +949,965 @@ mod tests {
+///
+/// Item must be defined as
+/// ```ignore
+/// #[pallet::config]
+/// pub trait Config: frame_system::Config + $optionally_some_other_supertraits
+/// $optional_where_clause
+/// {
+/// ...
+/// }
+/// ```
+/// I.e. a regular trait definition named `Config`, with supertrait `frame_system::Config`,
+/// optionally other supertrait and where clause.
+///
+/// The associated type `Event` is reserved, if defined it must bounds `From<Event>` and
+/// `IsType<<Self as frame_system::Config>::Event>`, see `#[pallet::event]` for more information.
+///
+/// To put `Get` associated type into metadatas, use the attribute `#[pallet::constant]`, e.g.:
+/// ```ignore
+/// #[pallet::config]
+/// pub trait Config: frame_system::Config {
+///		#[pallet::constant]
+///		type Foo: Get<u32>;
+/// }
+/// ```
+///
+/// To bypass the `frame_system::Config` supertrait check, use the attribute
+/// `#[pallet::disable_frame_system_supertrait_check]`, e.g.:
+/// ```ignore
+/// #[pallet::config]
+/// #[pallet::disable_frame_system_supertrait_check]
+/// pub trait Config: pallet_timestamp::Config {}
+/// ```
+///
+/// ### Macro expansion:
+///
+/// The macro expand pallet constant metadata with the information given by `#[pallet::constant]`.
+///
+/// # Pallet struct placeholder: `#[pallet::pallet]` mandatory
+///
+/// The placeholder struct, on which is implemented pallet informations.
+///
+/// Item must be defined as followed:
+/// ```ignore
+/// #[pallet::pallet]
+/// pub struct Pallet<T>(PhantomData<T>);
+/// ```
+/// I.e. a regular struct definition named `Pallet`, with generic T and no where clause.
+///
+/// To generate a `Store` trait associating all storages, use the attribute
+/// `#[pallet::generate_store($vis trait Store)]`, e.g.:
+/// ```ignore
+/// #[pallet::pallet]
+/// #[pallet::generate_store(pub(super) trait Store)]
+/// pub struct Pallet<T>(PhantomData<T>);
+/// ```
+/// More precisely the store trait contains an associated type for each storage. It is implemented
+/// for `Pallet` allowing to access the storage from pallet struct.
+///
+/// ### Macro expansion:
+///
+/// The macro add this attribute to the struct definition:
+/// ```ignore
+/// #[derive(
+/// 	frame_support::CloneNoBound,
+/// 	frame_support::EqNoBound,
+/// 	frame_support::PartialEqNoBound,
+/// 	frame_support::RuntimeDebugNoBound,
+/// )]
+/// ```
+///
+/// It implements on pallet:
+/// * [`GetPalletVersion`]
+/// * [`OnGenesis`]: contains some logic to write pallet version into storage.
+/// * [`ModuleErrorMetadata`]: using error declared or no metadata.
+///
+/// If attribute generate_store then macro create the trait `Store` and implement it on `Pallet`.
+///
+/// # Interface: `#[pallet::interface]` mandatory
+///
+/// Implementation of `Interface` on `Pallet` allowing to define some specific pallet logic.
+///
+/// Item must be defined as
+/// ```ignore
+/// #[pallet::interface]
+/// impl<T: Config> Interface<BlockNumberFor<T>> for Pallet<T> $optional_where_clause {
```

Comment:
Have we bikeshedded that name yet? :-P
`on_initialize` and friends sound more like hooks or callbacks to me. "Interface" is a super generic name.
What do you think?

Reply:
`Hooks` seems better to me indeed, I feel a bit less convinced by `Callbacks` but I'm definitly ok for renames